drivers/include/usb/usbdev.h

```diff
@@ -0,0 +1,337 @@
+ * had an interrupt that needs to be serviced in a non-interrupt context. This
+ * requires the usbdev_driver::esr function to be called.
+ *
+ * @file
+ * @brief       Definitions low-level USB driver interface
+ *
+ * @author      Koen Zandberg <koen@bergzand.net>
+ */
+
+#ifndef USB_USBDEV_H
+#define USB_USBDEV_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "usb.h"
+#include "usb/usbopt.h"
+
+typedef struct usbdev usbdev_t;
+typedef struct usbdev_ep usbdev_ep_t;
+
+/**
+ * Statically allocated buffer space for endpoints.
+ *
+ * When the device doesn't have dedicated memory for endpoint buffers, a
+ * buffer of this size is allocated to contain the endpoint buffers. Only
+ * needs to be as big as the total buffer space required by all endpoints
+ */
+#ifndef USBDEV_EP_BUF_SPACE
+#define USBDEV_EP_BUF_SPACE     1024
+#endif
+
+/**
+ * Number of USB IN and OUT endpoints allocated
+ *
+ * Configures the number of endpoints allocated. An equal number of IN and OUT
+ * endpoints are allocated
+ */
+#ifndef USBDEV_NUM_ENDPOINTS
+#define USBDEV_NUM_ENDPOINTS       8
+#endif
+
+
+/**
+ * @brief   Possible event types that are send from the device driver to the
+ *          upper layer
+ */
+typedef enum {
+    /**
+     * @brief Driver needs it's ISR handled
+     */
+    USBDEV_EVENT_ESR = 1,
+
+    /**
+     * @brief Line reset occured
+     *
+     * A line reset is a host initiated USB reset to the peripheral
+     *
+     * The peripheral driver must clears the following settings before
+     * emitting this event:
+     *  - Device address
+     *  - Endpoint configuration, including stall settings
+     */
+    USBDEV_EVENT_RESET,
+
+    /**
+     * @brief Transaction completed event.
+     *
+     * An endpoint must emit this event after a transaction with the host
+     * occured to indicate that the data in the buffer is used or new
+     * depending on the endpoint direction
+     */
+    USBDEV_EVENT_TR_COMPLETE,
+
+    /**
+     * @brief Transaction stall event.
+     *
+     * An endpoint should emit this event after a stall reply has been
+     * transmitted to the host
+     */
+    USBDEV_EVENT_TR_STALL,
+
+    /**
+     * @brief Transaction fail event.
+     *
+     * An endpoint should emit this event after a nack reply to the host
+     */
+    USBDEV_EVENT_TR_FAIL,
+    /* expand list if required */
+} usbdev_event_t;
+
+/**
+ * @brief   Event callback for signaling usbdev event to upper layers
+ *
+ * @param[in] usbdev        usbdev context
+ * @param[in] event         type of the event
+ */
+typedef void (*usbdev_event_cb_t)(usbdev_t *usbdev, usbdev_event_t event);
+
+/**
+ * @brief usbdev device descriptor
+ */
+struct usbdev {
+    const struct usbdev_driver *driver;     /**< usbdev driver struct   */
+    usbdev_event_cb_t cb;                   /**< Event callback supplied by
+                                              *  upper layer */
```

Comment:
What about adding `UsbDevice` struct from atmel vendor files ? In the current state, this driver works fine with one USB device instance but maybe one day we will have several USB within a MCU.
This way we can replace all USB->DEVICE from the driver by a more generic call.

Reply:
Agreed

Reply:
I assume you want to add that to the `sam0_common_usb_t` and not to the generic `usbdev` struct.

I think your point is valid, but I have to check if the current API allows this. If it doesn't, I would assume this to be a shortcomming of the API.

Reply:
> I assume you want to add that to the `sam0_common_usb_t` and not to the generic `usbdev` struct.

Yes

Reply:
Done, I'm not fully content with the current implementation for retrieving the context in the isr, but this can be reworked as soon as we actually have a device with multiple usb peripherals